### DISCUSSION TITLE
Use case for "large-dates" feature flag

### DISCUSSION THREAD

**Discussion by jhpratt** [time-rs/time#306] [Seeking input]
This discussion is for collecting the various use cases of having dates outside of the range -9999-01-01 to +9999-12-31. In time 0.3, support for these values is placed behind the "large-dates" flag, which allows some optimizations for what I believe to be the most common use case.

As indicated in the crate's documentation, this flag may be removed in a future release if there is insufficient demand. Leave your use case below, and use the :+1: reaction if you share this use case. Please do not :+1: a use case that you do not actually need this feature for. Off-topic comments will likely be outright deleted.

**Comment by poliorcetics** [comment]
We need the feature to safely interact with Windows APIs: [`FILETIME`](https://docs.microsoft.com/en-us/windows/win32/api/minwinbase/ns-minwinbase-filetime) and [`SYSTEMTIME`](https://docs.microsoft.com/en-us/windows/win32/api/minwinbase/ns-minwinbase-systemtime) can both go way further than +9999-12-31 ([around 30827](http://stackoverflow.com/questions/9999393/ddg#18188484) in fact) so this feature is needed when converting times we got from the Windows API.

**Reply by jhpratt** [reply]
For what purpose do you need to set/get dates from files that are thousands of years in the future?

**Reply by poliorcetics** [reply]
Ideally we wouldn't, but we are writing a security product and failing to get some information from a suspicious file because we couldn't parse a date in the Windows registry is not acceptable for us.

I understand the parsing ambiguities introduced by this feature though, and I can participate in discussions on how to resolve or minimize them if you want, just ping me on the relevant discussions/issues.

**Reply by jhpratt** [reply]
After a bit of thinking today, I believe it may be possible to get the best of both worlds. For parsing, a new modifier could be introduced to indicate large dates (those with 5 or more digits in the year) can be parsed. This would let the person parsing the string to choose.

The optimizations that are performed when the flag is _not_ enabled just involve using a smaller integer type if I remember correctly. This could be gated to only be performed when the year is in a certain range, which should largely preserve the performance gain when possible.

I'll need to think about this more, but if I'm correct that it's feasible then the feature gate could be eliminated in a breaking release and just become the default behavior. It's not possible to do this in a non-breaking release, as the documentation says the range is ±9999 years without the flag enabled.

**Reply by poliorcetics** [reply]
Yes, a different modifier for parsing would solve pretty much everything (the `year` modifier would probably need to voluntarily stop at 4 digits though, to avoid breaking too much code ?).

The smaller integer type is not that much of an issue I think, but I'm not an embedded/low constraint user so I wouldn't know about their opinion, though anything less than a u32 is probably handled as a u32 by most embedded CPUs IIRC.

**Reply by jhpratt** [reply]
Modifier = the thing after `year`, such as `padding:space`. The _component_ is `year`. So you could have something like `[year large:true]` (it won't be exactly that). The default would be to have a maximum of four digits.

The smaller integer type isn't really a constraint thing, it's more a performance improvement on the order of 15-20% for certain methods.

**Reply by poliorcetics** [reply]
That's a significant hit for those methods if we move to a larger integer type, what are the methods concerned ?

**Reply by jhpratt** [reply]
I'd have to check to be certain, but I believe it involved calculating the Julian day number (either from or to, I forget). That seems inconsequential, but it's actually the underlying manner in which all arithmetic is done with `Date` objects. That's why I said I think I can branch when the value is in the requisite range for the optimization.

**Reply by poliorcetics** [reply]
If you think this can be done without impacting everyone else by 15%, sure, let's do it

**Reply by jhpratt** [reply]
At some point I can run the relevant benchmarks when it's behind an `if` statement. "At some point" is probably a couple weeks away at least, as I'm writing an RFC, need to review PRs for this repo, and want to do some other unrelated things.

**Reply by poliorcetics** [reply]
This is not urgent, don't worry about it, I just wanted to chime in to avoid the large dates feature disappearing without replacement. If you have guidance instructions somewhere I can have a look to see if I can work on it, but don't do it if writing them would take too long !

Good luck on the RFC, I should really publish mine someday too

**Reply by jhpratt** [reply]
It's a breaking change as I mentioned earlier, so it _definitely_ isn't urgent. You're (obviously) the first one to even mention having a use case.

**Comment by ChristianBeilschmidt** [comment]
We do geoprocessing and usually process data that is collected in recent centuries and now. But there are scenarios, e.g. in landscape modeling, where we look at historic dates way beyond -1k. These are quite inaccurate, but at least with `large-dates` we can use this library for that and don't need to specially treat this data.

We are currently switching from `chrono` to `time` and `chrono` had wider min and max values. So we are thinking about activating this feature and it would be sad to see it gone.

**Comment by ColinFinck** [comment]
Thanks for having this discussion before making breaking changes!

Just like @poliorcetics, I'm targeting the Windows world, which internally stores time in an unsigned 64-bit integer that counts the number of 100-nanosecond intervals since January 1st, 1601. My [`ntfs` crate](https://github.com/ColinFinck/ntfs) implements the internals of the NTFS filesystem, which also include this time format.

To interoperate with other parts of the Rust ecosystem, I've just implemented converters from and to `OffsetDateTime` of the `time` crate. Particularly, a conversion from my `NtfsTime` to `OffsetDateTime` is infallible, because `OffsetDateTime`'s nanosecond-precision `i128` is a superset of `NtfsTime`'s 100-nanosecond-precision `u64`.
However, this also requires full support in `time` for years up to the maximum 60056 supported by `NtfsTime`. The crate must not impose arbitrary limitations on the value range of the internal `i128`.
In fact, I'm testing my converters with year 60056 dates to verify their correctness: https://github.com/ColinFinck/ntfs/blob/d54266330757ad2da3da38cd17ff1445a9cee4ae/src/time.rs#L111-L115

**Reply by jhpratt** [reply]
Thanks. I should actually update the documentation to reflect the fact that the behavior isn't going away in 0.4. There will be a slight change in behavior, but that should make it simpler for everybody.
